src/core/border_router/routing_manager.cpp

```diff
@@ -0,0 +1,740 @@
+}
+
+void RoutingManager::RecvIcmp6Message(uint32_t            aInfraIfIndex,
+                                      const Ip6::Address &aSrcAddress,
+                                      const uint8_t *     aBuffer,
+                                      uint16_t            aBufferLength)
+{
+    const Ip6::Icmp::Header *icmp6Header;
+
+    // The Border Routing function is enabled only when we are attached.
+    VerifyOrExit(IsInitialized() && Get<Mle::MleRouter>().IsAttached());
+
+    VerifyOrExit(aInfraIfIndex == mInfraIfIndex);
+    VerifyOrExit(aBuffer != nullptr && aBufferLength >= sizeof(*icmp6Header));
+
+    icmp6Header = reinterpret_cast<const Ip6::Icmp::Header *>(aBuffer);
+
+    switch (icmp6Header->GetType())
+    {
+    case Ip6::Icmp::Header::kTypeRouterAdvert:
+        HandleRouterAdvertisement(aSrcAddress, aBuffer, aBufferLength);
+        break;
+    case Ip6::Icmp::Header::kTypeRouterSolicit:
+        HandleRouterSolicit(aSrcAddress, aBuffer, aBufferLength);
+        break;
+    default:
+        ExitNow();
+    }
+
+exit:
+    return;
+}
+
+void RoutingManager::HandleNotifierEvents(Events aEvents)
+{
+    VerifyOrExit(IsInitialized());
+
+    if (aEvents.Contains(kEventThreadRoleChanged))
+    {
+        if (Get<Mle::MleRouter>().IsAttached())
+        {
+            Start();
+        }
+        else
+        {
+            Stop();
+        }
+    }
+
+    if (aEvents.Contains(kEventThreadNetdataChanged))
+    {
+        if (Get<Mle::MleRouter>().IsAttached())
+        {
+            EvaluateRoutingPolicy();
+        }
+    }
+
+exit:
+    return;
+}
+
+uint8_t RoutingManager::EvaluateOmrPrefix(Ip6::Prefix *aNewOmrPrefixes)
+{
+    uint8_t                         newOmrPrefixesNum = 0;
+    NetworkData::Iterator           iterator          = NetworkData::kIteratorInit;
+    NetworkData::OnMeshPrefixConfig onMeshPrefixConfig;
+
+    VerifyOrExit(Get<Mle::MleRouter>().IsAttached());
+
+    while (Get<NetworkData::Leader>().GetNextOnMeshPrefix(iterator, onMeshPrefixConfig) == OT_ERROR_NONE)
+    {
+        if (!IsValidOmrPrefix(onMeshPrefixConfig.GetPrefix()) || !onMeshPrefixConfig.mDefaultRoute ||
+            !onMeshPrefixConfig.mSlaac || onMeshPrefixConfig.mDp)
+        {
+            continue;
+        }
+
+        if (newOmrPrefixesNum < kMaxOmrPrefixNumber)
+        {
+            aNewOmrPrefixes[newOmrPrefixesNum++] = onMeshPrefixConfig.GetPrefix();
+        }
+        else
+        {
+            otLogInfoBr("EvaluateOmrPrefix: too many OMR prefixes, ignoring prefix %s",
+                        onMeshPrefixConfig.GetPrefix().ToString().AsCString());
+        }
+    }
+
+    if (newOmrPrefixesNum == 0)
+    {
+        if (!IsLocalOmrPrefixAdvertised())
+        {
+            otLogInfoBr("EvaluateOmrPrefix: no valid OMR prefixes found in Thread network");
+            if (PublishLocalOmrPrefix() == OT_ERROR_NONE)
+            {
+                aNewOmrPrefixes[newOmrPrefixesNum++] = mLocalOmrPrefix;
+            }
```

Comment:
If `PublishLocalOmrPrefix` failed here. how do we make sure that `EvaluateRoutingPolicy` will be executed again in the near future. 

There are 3 sources that can trigger `EvaluateRoutingPolicy`:
- Router solicit timer fired for the 3rd time
- RA received with PIO of an On-Link prefix
- Thread network data changed

None of these conditions can we guarantee that will happen in the future. 

My suggestion would be using a periodical timer to trigger `EvaluateRoutingPolicy` timely so that  any abnormality can be fixed eventually.

Reply:
handled by start the router advertisement timer to re-evaluate the OMR prefix in the future.